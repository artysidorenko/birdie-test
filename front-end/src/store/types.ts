// #1 INDIVIDUAL DATAPOINT INTERFACES

export interface Event {
  id: string;
  event_type: string;
  timestamp: Date;
  caregiver_id: string;
  visit_id: string;
  care_recipient_id: string;
  note: string;
}

export interface FluidIntake extends Event {
  fluid: string;
  consumed_volume_ml: number;
}

export interface FoodIntake extends Event {
  meal: string;
  note: string;
}

// includes general, mental health & physical observations
export interface Observation extends Event {
  note: string;
}

export interface PadCondition extends Event {
  pad_condition: string;
}

export interface Mood extends Event {
  mood: string;
}

export interface Medication extends Event {
  medication_taken: boolean;
  medication_failure_reason: string | undefined | null;
}

export interface Task extends Event {
  task_definition_description: string;
  task_schedule_note: string;
}

// #2 ACTION TYPES AND INTERFACES

// #2 ACTION TYPES AND INTERFACES
export enum actions {
  GET_EVENTS = 'GET_EVENTS',
  GET_FLUID_INTAKE = 'GET_FLUID_INTAKE',
  GET_FOOD_INTAKE = 'GET_FOOD_INTAKE',
  GET_OBSERVATIONS = 'GET_OBSERVATIONS',
  GET_PAD_CONDITION = 'GET_PAD_CONDITION',
  GET_MOODS = 'GET_MOODS',
  GET_MEDICATION = 'GET_MEDICATION',
  GET_TASKS = 'GET_TASKS',
  GET_EVENTS_RESPONSE = 'GET_EVENTS_RESPONSE ',
  GET_FLUID_INTAKE_RESPONSE = 'GET_FLUID_INTAKE_RESPONSE ',
  GET_FOOD_INTAKE_RESPONSE = 'GET_FOOD_INTAKE_RESPONSE ',
  GET_OBSERVATIONS_RESPONSE = 'GET_OBSERVATIONS_RESPONSE ',
  GET_PAD_CONDITION_RESPONSE = 'GET_PAD_CONDITION_RESPONSE ',
  GET_MOODS_RESPONSE = 'GET_MOODS_RESPONSE ',
  GET_MEDICATION_RESPONSE = 'GET_MEDICATION_RESPONSE ',
  GET_TASKS_RESPONSE = 'GET_TASKS_RESPONSE',
  SET_ID = 'SET_ID'
}

export enum status {
  LOADING = 'loading',
  SUCCESS = 'success',
  FAILURE = 'failure'
}

export interface SetID {
  type: typeof actions.SET_ID;
  id: string;
}

export interface GetEvents {
  type: typeof actions.GET_EVENTS;
  id: string;
}
export interface GetEventsResponse {
  type: typeof actions.GET_EVENTS_RESPONSE;
  status: status;
  response: Event[] | Error;
}

export interface GetFluidIntake {
  type: typeof actions.GET_FLUID_INTAKE;
  id: string;
}
export interface GetFluidIntakeResponse {
  type: typeof actions.GET_FLUID_INTAKE_RESPONSE;
  status: status;
  response: FluidIntake[] | Error;
}

export interface GetFoodIntake {
  type: typeof actions.GET_FOOD_INTAKE;
  id: string;
}
export interface GetFoodIntakeResponse {
  type: typeof actions.GET_FOOD_INTAKE_RESPONSE;
  status: status;
  response: FoodIntake[] | Error;
}

export interface GetObservations {
  type: typeof actions.GET_OBSERVATIONS;
  id: string;
}
export interface GetObservationsResponse {
  type: typeof actions.GET_OBSERVATIONS_RESPONSE;
  status: status;
  response: Observation[] | Error;
}

export interface GetPadCondition {
  type: typeof actions.GET_PAD_CONDITION;
  id: string;
}
export interface GetPadConditionResponse {
  type: typeof actions.GET_PAD_CONDITION_RESPONSE;
  status: status;
  response: PadCondition[] | Error;
}

export interface GetMoods {
  type: typeof actions.GET_MOODS;
  id: string;
}
export interface GetMoodsResponse {
  type: typeof actions.GET_MOODS_RESPONSE;
  status: status;
  response: Mood[] | Error;
}

export interface GetMedication {
  type: typeof actions.GET_MEDICATION;
  id: string;
}
export interface GetMedicationResponse {
  type: typeof actions.GET_MEDICATION_RESPONSE;
  status: status;
  response: Medication[] | Error;
}

export interface GetTasks {
  type: typeof actions.GET_TASKS;
  id: string;
}
export interface GetTasksResponse {
  type: typeof actions.GET_TASKS_RESPONSE;
  status: status;
  response: Task[] | Error;
}

export type ActionTypes =
  | SetID
  | GetEvents
  | GetFluidIntake
  | GetFoodIntake
  | GetMedication
  | GetMoods
  | GetObservations
  | GetPadCondition
  | GetTasks
  | GetEventsResponse
  | GetFluidIntakeResponse
  | GetFoodIntakeResponse
  | GetMedicationResponse
  | GetMoodsResponse
  | GetObservationsResponse
  | GetPadConditionResponse
  | GetTasksResponse;

// #3 STATE TYPES

export type LoadingState = {
  readonly status: status | null | undefined;
  readonly error: string | null | undefined;
};

export type EventsState = LoadingState & {
  readonly events: Event[];
};
export type FluidState = LoadingState & {
  readonly fluid_intake: FluidIntake[];
};
export type FoodState = LoadingState & {
  readonly food_intake: FoodIntake[];
};
export type ObservationState = LoadingState & {
  readonly observations: Observation[];
};
export type PadState = LoadingState & {
  readonly pad_condition: PadCondition[];
};
export type MoodState = LoadingState & {
  readonly moods: Mood[];
};
export type MedicationState = LoadingState & {
  readonly medication: Medication[];
};
export type TaskState = LoadingState & {
  readonly tasks: Task[];
};

export type DataState = {
  readonly id: string,
  readonly events: EventsState;
  readonly fluid_intake: FluidState;
  readonly food_intake: FoodState;
  readonly observations: ObservationState;
  readonly pad_condition: PadState;
  readonly moods: MoodState;
  readonly medication: MedicationState;
  readonly tasks: TaskState;
};