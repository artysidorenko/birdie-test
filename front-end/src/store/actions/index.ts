import { ActionTypes, actions } from '../types';

export function getEvents(): ActionTypes {
  return {
    type: actions.GET_EVENTS
  };
}
export function getFluidIntake(): ActionTypes {
  return {
    type: actions.GET_FLUID_INTAKE
  };
}
export function getFoodIntake(): ActionTypes {
  return {
    type: actions.GET_FOOD_INTAKE
  };
}
export function getMedication(): ActionTypes {
  return {
    type: actions.GET_MEDICATION
  };
}
export function getMoods(): ActionTypes {
  return {
    type: actions.GET_MOODS
  };
}
export function getObservations(): ActionTypes {
  return {
    type: actions.GET_OBSERVATIONS
  };
}
export function getPadCondition(): ActionTypes {
  return {
    type: actions.GET_PAD_CONDITION
  };
}
export function getTasks(): ActionTypes {
  return {
    type: actions.GET_TASKS
  };
}