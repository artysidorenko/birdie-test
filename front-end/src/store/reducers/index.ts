import { combineReducers, Reducer } from 'redux';
import { actions, DataState } from '../types';
import { initialState } from './initialState';

export type RootState = Readonly<{
  readonly data: DataState
}>;

export const dataReducer: Reducer<DataState> = (state = initialState, action) => {
  switch (action.type) {
    case actions.SET_ID:
      return {
        ...state,
        id: action.id
      };
    case actions.GET_EVENTS_RESPONSE:
      return {
        ...state,
        events: {
          ...state.events,
          status: action.status,
          ...(action.status === 'failure' && {error: action.response}),
          ...(action.status === 'success' && {events: action.response})
        }
      };
    case actions.GET_FLUID_INTAKE_RESPONSE:
      return {
        ...state,
        fluid_intake: {
          ...state.fluid_intake,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { fluid_intake: action.response })
        }
      };
    case actions.GET_FOOD_INTAKE_RESPONSE:
      return {
        ...state,
        food_intake: {
          ...state.food_intake,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { food_intake: action.response })
        }
      };
    case actions.GET_MEDICATION_RESPONSE:
      return {
        ...state,
        medication: {
          ...state.medication,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { medication: action.response })
        }
      };
    case actions.GET_MOODS_RESPONSE:
      return {
        ...state,
        moods: {
          ...state.moods,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { moods: action.response })
        }
      };
    case actions.GET_OBSERVATIONS_RESPONSE:
      return {
        ...state,
        observations: {
          ...state.observations,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { observations: action.response })
        }
      };
    case actions.GET_PAD_CONDITION_RESPONSE:
      return {
        ...state,
        pad_condition: {
          ...state.pad_condition,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { pad_condition: action.response })
        }
      };
    case actions.GET_TASKS_RESPONSE:
      return {
        ...state,
        tasks: {
          ...state.tasks,
          status: action.status,
          ...(action.status === 'failure' && { error: action.response }),
          ...(action.status === 'success' && { tasks: action.response })
        }
      };
    default:
      return state;
  }
};

export const rootReducer = combineReducers<RootState>({
  data: dataReducer
});