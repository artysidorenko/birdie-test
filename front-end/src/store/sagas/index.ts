import { all, spawn } from 'redux-saga/effects';

import dataSaga from './dataSaga';

function* initSaga() {
  yield all([spawn(dataSaga)]);
}

export default initSaga;