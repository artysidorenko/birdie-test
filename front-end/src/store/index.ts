import { applyMiddleware, compose, createStore, GenericStoreEnhancer } from 'redux';
import createSagaMiddleware from 'redux-saga';
import * as _history from 'history';
import { rootReducer, RootState } from '@App/store/reducers';
import initSaga from '@App/store/sagas';
import { DataState } from './types';

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION__: () => undefined;
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__: (arg: GenericStoreEnhancer) => undefined;
  }
}

const sagaMiddleware = createSagaMiddleware();
export const history = _history.createBrowserHistory();

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(sagaMiddleware)),
);

sagaMiddleware.run(initSaga);

export default store;

export const mapStateToProps = (state: RootState, ownProps: object): DataState => ({
  events: state.data.events,
  fluid_intake: state.data.fluid_intake,
  food_intake: state.data.food_intake,
  observations: state.data.observations,
  pad_condition: state.data.pad_condition,
  moods: state.data.moods,
  medication: state.data.medication,
  tasks: state.data.tasks
});