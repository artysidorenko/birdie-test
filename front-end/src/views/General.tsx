import * as React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import styled from 'styled-components';

import { RootState } from '@App/store/reducers';
import { mapStateToProps } from '../store';
import { DataState, status as fetchStatus } from '../store/types';
import { getEvents, getMoods } from '../store/actions';

import VisitsBar from '@App/components/graphs/VisitsBar';
import CarerPie from '@App/components/graphs/CarerPie';
import MoodsBar from '@App/components/graphs/MoodsBar';

interface ActionProps {
  getEvents: () => void;
  getMoods: () => void;
}

type AppProps = DataState & ActionProps;

const cardFormat = `
  box-shadow: 0 0.46875rem 2.1875rem rgba(90,97,105,.1),
    0 0.9375rem 1.40625rem rgba(90,97,105,.1),
    0 0.25rem 0.53125rem rgba(90,97,105,.12),
    0 0.125rem 0.1875rem rgba(90,97,105,.1);

  padding: 10px;
  margin-bottom: 20px;
  `;

const StyledWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-evenly;

  @media screen and (max-width: 950px) {
    flex-direction: column;
    align-items: center;
  }
`;

const StyledContainer = styled.div`
  position: relative;
  height: 40vh;
  width: 82.5vw;
  ${cardFormat}
`;
const StyledContainerSmall = styled.div`
  position: relative;
  height: 40vh;
  width: 40vw;
  margin-left: 10px;
  margin-right: 10px;
  ${cardFormat}

  @media screen and (max-width: 950px) {
    width: 80vw;
    margin-left: 0px;
    margin-right: 0px;
  }
`;
const Row = styled.div`
  position: relative;
  display: flex;
  justify-content: space-evenly;

  @media screen and (max-width: 950px) {
    flex-direction: column;
    align-items: center;
  }
`;

class General extends React.Component<AppProps> {
  displayName: '';

  componentDidMount() {
    this.props.getEvents();
    this.props.getMoods();
  }

  render() {
    const { status: eventStatus, events } = this.props.events;
    const { status: moodStatus, moods } = this.props.moods;

    return (
      <StyledWrapper>
        {eventStatus === fetchStatus.SUCCESS &&
        moodStatus === fetchStatus.SUCCESS ? (
          <React.Fragment>
            <Row>
              <StyledContainerSmall id="mood">
                <MoodsBar data={moods} />
              </StyledContainerSmall>
              <StyledContainerSmall id="carers">
                <CarerPie data={events} />
              </StyledContainerSmall>
            </Row>
            <StyledContainer id="visits">
              <VisitsBar data={events} />
            </StyledContainer>
          </React.Fragment>
        ) : (
          <span>Loading...</span>
        )}
      </StyledWrapper>
    );
  }
}

const mapDispatchToProps = (dispatch: Dispatch<RootState>) => ({
  getEvents: () => dispatch(getEvents()),
  getMoods: () => dispatch(getMoods())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(General);