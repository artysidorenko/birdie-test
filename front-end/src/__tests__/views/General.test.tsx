import * as React from 'react';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import { expect as _expect } from 'chai';

import General from '@App/views/General';
import store from '@App/store';

/* Normally part of src/setupTests.js but ... */
import { configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
/* ... file was not being picked up by CRA so added manually */

/* tslint:disable:no-unused-expression */

describe('General Page', () => {

  it('renders without crashing (shallow)', () => {
    const wrapper = shallow(<Provider store={store}><General /></Provider>);
    expect(wrapper).toMatchSnapshot();
  });
});
