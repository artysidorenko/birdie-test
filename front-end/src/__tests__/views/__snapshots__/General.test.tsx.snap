// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General Page renders without crashing (shallow) 1`] = `<Connect(General) />`;
